MultiServer.py

```diff
@@ -634,13 +634,30 @@ def get_hint_cost(self, slot):
             return max(1, int(self.hint_cost * 0.01 * len(self.locations[slot])))
         return 0
 
-    def recheck_hints(self, team: typing.Optional[int] = None, slot: typing.Optional[int] = None):
+    def recheck_hints(self, team: typing.Optional[int] = None, slot: typing.Optional[int] = None,
+                      changed: typing.Optional[typing.Set[team_slot]] = None,
+                      already_updated: typing.Optional[typing.Set[team_slot]] = None) -> None:
+        if not already_updated:
+            already_updated = set()
+        needs_updating = set()
         for hint_team, hint_slot in self.hints:
+            if (hint_team, hint_slot) in already_updated:
+                continue
             if (team is None or team == hint_team) and (slot is None or slot == hint_slot):
-                self.hints[hint_team, hint_slot] = {
-                    hint.re_check(self, hint_team) for hint in
-                    self.hints[hint_team, hint_slot]
-                }
+                new_hints: typing.Set[NetUtils.Hint] = set()
+                for hint in self.hints[hint_team, hint_slot]:
+                    new_hint = hint.re_check(self, hint_team)
+                    new_hints.add(new_hint)
+                    if hint != new_hint:
```

Comment:
Nitpicking: Wouldn't it be better to invert the condition and use a `continue` to have less indentation? It makes the method more readable IMO 

(same for line 646)